cub/test/catch2_test_util_device.cu

```diff
@@ -87,3 +87,75 @@ CUB_TEST("CUB correctly identifies the ptx version the kernel was compiled for",
   REQUIRE(ptx_version == kernel_cuda_arch);
   REQUIRE(host_ptx_version == kernel_cuda_arch);
 }
+
+#ifdef __CUDA_ARCH_LIST__
+CUB_TEST("PtxVersion returns a value from __CUDA_ARCH_LIST__", "[util][dispatch]")
+{
+  int ptx_version = 0;
+  cub::PtxVersion(ptx_version);
+  const auto arch_list = std::vector<int>{__CUDA_ARCH_LIST__};
+  REQUIRE(std::find(arch_list.begin(), arch_list.end(), ptx_version) != arch_list.end());
+}
+#endif
+
+#ifdef __CUDA_ARCH_LIST__
+// We list policies for all virtual architectures that __CUDA_ARCH_LIST__ can contain, so the actual architectures the
+// tests are compiled for should match to one of those
+struct policy_hub
+{
+#  define GEN_POLICY(cur, prev)                                             \
+    struct policy##cur : cub::ChainedPolicy<cur, policy##cur, policy##prev> \
+    {                                                                       \
+      static constexpr int value = cur;                                     \
+    }
+  // for the list of supported architectures, see libcudacxx/include/nv/target
+  GEN_POLICY(350, 350);
+  GEN_POLICY(370, 350);
+  GEN_POLICY(500, 370);
+  GEN_POLICY(520, 500);
+  GEN_POLICY(530, 520);
+  GEN_POLICY(600, 530);
+  GEN_POLICY(610, 600);
+  GEN_POLICY(620, 610);
+  GEN_POLICY(700, 620);
+  GEN_POLICY(720, 700);
+  GEN_POLICY(750, 720);
+  GEN_POLICY(800, 750);
+  GEN_POLICY(860, 800);
+  GEN_POLICY(870, 860);
+  GEN_POLICY(890, 870);
+  GEN_POLICY(900, 890);
+  GEN_POLICY(1000, 900);
+  // add more policies here when new architectures emerge
+  GEN_POLICY(2000, 1000); // non-existing architecture, just to test pruning
+#  undef GEN_POLICY
+
+  using max_policy = policy2000;
+};
+
+// Check that selected is one of arches
+template <int selected, int... arch_list>
+struct check
+{
+  static_assert(::cuda::std::_Or<::cuda::std::bool_constant<selected == arch_list>...>::value, "");
+  using type = cudaError_t;
+};
+
+struct Closure
+{
+  // We need to fail template instantiation if ActivePolicy::value is not one from the __CUDA_ARCH_LIST__
+  template <typename ActivePolicy>
+  _CCCL_HOST_DEVICE auto Invoke() const -> typename check<ActivePolicy::value, __CUDA_ARCH_LIST__>::type
+  {
+    return cudaSuccess;
+  }
+};
+
+CUB_TEST("ChainedPolicy prunes based on __CUDA_ARCH_LIST__", "[util][dispatch]")
```

Comment:
important: Currently, we are only testing that the `Closure::Invoke()` template is getting _instantiated_ only for architectures that are in `__CUDA_ARCH_LIST__`. This verifies that the _optimization_ of pruning itself works as expected. 

I think it is important to also add tests that verify that `Invoke` gets _called_ for the `ActivePolicy` that we'd expect it to get invoked for. That is, the template instantiation with `ActivePolicyT` being the closest policy from the chain that is equal to or larger than the given `device_ptx_version`. This is essential for the correctness of our algorithms. Because if there is a discrepancy between the `Invoke` template instantiation that actually gets called (at run-time, to determine the launch configuration) and the kernel querying the policy applicable during a the compilation pass of a certain architecture (using static querying at compilation time, using `ChainedPolicy::ActivePolicy` type alias) the algorithms will not work correctly. Since this mechanism is at the core of pretty much anything in CUB, we want to make sure all works correctly.

I think this could be achieved adding a `device_ptx_version` member to the `Closure` and set that member to the value retrieved from `cub::PtxVersion()`. `Closure::Invoke()` would then need to verify that (a) `ActivePolicy::value >= device_ptx_version` and that the next higher policy (actually _all_ "higher" policies) in the chain didn't qualify (i.e., its `value` is greater than `device_ptx_version`). We will then return `cudaSuccess` if ok, and something like `cudaErrorInvalidValue` if that's not the case.

Reply:
For that purpose, we should also have a test that has wide gaps in the tuning policy chain, possibly just define a non-existing high arch and `350`.

Reply:
Those are great suggestions, I added those tests!